### ISSUE TITLE
[FFmpeg] segments dropped when recording HLS

### ISSUE THREAD

**Issue by soimort** [soimort/you-get#1270]
This could be a bug (misfeature?) in `ffmpeg`, but we should really work around this:

```
skipping 1 segments ahead, expired from playlists
```

Skipped segments can cause periodic freezes in the produced video (it seems audio track is unaffected, though). In most cases when one *downloads* a live stream, it is desirable to have all segments (that make up the complete video) rather than performing real-time.

**Comment by cnbeining** [comment]
So this is my point: only use ffmpeg when we have to.

2016年7月10日星期日，Mort the Bot <notifications@github.com> 写道：

**Comment by lamlam123tld3** [comment]
my suggestion:
download all the segments through reliable method,
change the m3u8 file pointing to local downloaded files
run ffmpeg

**Comment by cnbeining** [comment]
That would means a hug change of the current algorithm. I am not so sure
whether it is worth it ...

2016年7月14日星期四，lamlam123tld3 <notifications@github.com> 写道：

**Comment by soimort** [comment]
To sum up @lamlam123tld3 's suggestion:
Just don't let ffmpeg handle `m3u8` directly.

Sounds feasible to me.

**Comment by lamlam123tld3** [comment]
in case the m3u8 segments had been encryted by AES or etc.
i would be better to throw it into ffmpeg to tackle it

**Comment by soimort** [comment]
> in case the m3u8 segments had been encryted by AES or etc.

That's right. As far as I'm concerned, though -- encryption is not involved in the live streams mentioned in #1218 and #1233.

**Comment by cnbeining** [comment]
What I am thinking are:

HLS involves a M3U file whose content changes every 3-5 secs, containing each link.

YES we CAN use ```m3u``` package, but I am not 100% sure what are we going to do with the segments:

- Are we going to download every available segment as we find it? Or we wait till the user press Ctrl+C then try to download everything? The first approach would mean a potential huge change to the current structure of the programme; while we CANNOT know whether the old segments are still available for download as the server can delete them as defined in the protocol.

- What if some of the segments are broken? There's no guarantee that we are recording the stream from the very beginning, thus the timestamp of the segments we have when trying to merge could be broken, and IMHO this is very likely to be the case.

- If ```ffmpeg``` cannot download the segments due to some reason(network? I didn't read the code of ffmpeg, nor do I think I shall do it anyways), what can we do to improve this situation? 

- Only under very rare situations that people shall use ```AES``` with ```HLS``` AFAIK... as cool kids are using Adobe HDS(which is more painful as we do not have the right tool other than asking for PHP installed, sigh).

**Comment by soimort** [comment]
> The first approach would mean a potential huge change to the current structure of the programme; while we CANNOT know whether the old segments are still available for download as the server can delete them as defined in the protocol.

Do a best-effort retrieval. This can't be worse than `ffmpeg`.

Or does anyone know if there's an option to ask `ffmpeg` not to expire segments, in case of a suboptimal network status?

**Comment by soimort** [comment]
Interesting fact: `vlc` does a good job streaming these HLS m3u8 playlists. `ffmpeg`/`ffplay` freezes the video output when a segment expires, and `mpv` has an even more obvious a/v desynchronization problem.

For now this issue is observed only on SHOWROOM, but not on LINE-LIVE or AMESTAGE. This is tricky to me, because the playback in `vlc` is still not as smooth as the in-browser Flash player (I don't know why).

```
[00007f4030c03d68] httplive stream error: EXT-X-MEDIA-SEQUENCE gap in playlist (new=794, old=793)
```

**Comment by cyfex** [comment]
Any update on this issue? I am facing and annoying at this.
